### ISSUE TITLE
DISTINCT with LIMIT in subqueries drops valid results

### ISSUE THREAD

**Issue by vincentvialard** [neo4j/neo4j#13197]
I stumbled upon a weird behaviour of subqueries using `DISTINCT` and `LIMIT` where results are lost whenever `UNWIND` is used before the subquery, **even if it is not used by the subquery**.  

I reduced the query complexity to a minimum and demonstrate it with a public dataset.

**Neo4j Version:** 4.4.20 Enterprise (in Neo4j Desktop)
**Operating System:** macOs 13.3.1

### Steps to reproduce
1) Download the Reactome dump from https://reactome.org/download/current/reactome.graphdb.dump
2) Add the file to Neo4j Desktop
3) Create a new DBMS v4.4.20 from the dump
4) Start the DBMS
5) Open the Neo4j Browser and run the following queries:

Base query:
```
CALL {
  MATCH (n:DatabaseObject)-[]->() 
  WHERE n.schemaClass="TopLevelPathway" 
  RETURN DISTINCT n 
  ORDER BY n
  LIMIT 200
}
RETURN count(n) as count;
```

Unused UNWIND + base query:
```
UNWIND [1] as one
CALL {
  MATCH (n:DatabaseObject)-[]->() 
  WHERE n.schemaClass="TopLevelPathway" 
  RETURN DISTINCT n 
  ORDER BY n
  LIMIT 200
}
RETURN count(n) as count;
```

Unused UNWIND + base query with `n.dbId` ordering instead of `n`
```
UNWIND [1] as one
CALL {
  MATCH (n:DatabaseObject)-[]->() 
  WHERE n.schemaClass="TopLevelPathway" 
  RETURN DISTINCT n 
  ORDER BY n.dbId
  LIMIT 200
}
RETURN count(n) as count;
````

### Expected behavior
The 3 queries return 200

### Actual behavior
The first and third queries do return 200, the second 119

Profiling the queries 2 and 3 shows where it goes wrong:

<img width="1252" alt="profiles" src="https://github.com/neo4j/neo4j/assets/16098519/e93bc308-b154-47aa-90e1-75d2a1a7ce0f">

### Remark
The problem is reproducible neither with Neo4j Community 4.4.20 nor with Neo4j Enterprise 5.8.0 (after conversion of the database), but the query profiling still exhibits some differences. So I would not exclude the possibilty that the problem also may occur for some other query in these versions.

**Comment by JonasHolgersson** [comment]
@vincentvialard Thank you for reporting! I could reproduce the issue and it definitely seems to be something going on here. We will investigate further and come back with our findings.

Regards,
Jonas

**Comment by JonasHolgersson** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [DISTINCT with LIMIT in subqueries drops valid results](https://trello.com/c/QJe67Xwl/4635-distinct-with-limit-in-subqueries-drops-valid-results)

**Comment by JonasHolgersson** [comment]
@vincentvialard Thanks again for reporting, this has been fixed and should be included in the next 4.4-release.

Regards,
Jonas
